primitives/src/vstaging/executor_params.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Abstract execution environment parameter set.
+//!
+//! Parameter set is encoded as an opaque vector which structure depends on the execution
+//! environment itself (except for environment type/version which is always represented
+//! by the first element of the vector). Decoding to a usable semantics structure is
+//! done in `polkadot-node-core-pvf`.
+
+use crate::v2::{BlakeTwo256, HashT as _};
+use parity_scale_codec::{Decode, Encode};
+use polkadot_core_primitives::Hash;
+use scale_info::TypeInfo;
+use sp_std::{ops::Deref, vec, vec::Vec};
+
+/// Execution environment type
+#[derive(Clone, Copy, Debug, Encode, Decode, PartialEq, Eq, TypeInfo)]
+pub enum ExecutionEnvironment {
+	/// Generic Wasmtime executor
+	WasmtimeGeneric = 0,
+}
+
+/// Executor instantiation strategy
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo)]
+pub enum ExecInstantiationStrategy {
+	/// Pooling copy-on-write
+	PoolingCoW,
+	/// Recreate instance copy-on-write
+	RecreateCoW,
+	/// Pooling
+	Pooling,
+	/// Recreate instance
+	Recreate,
+}
+
+/// A single executor parameter
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo)]
+pub enum ExecutorParam {
+	/// ## General parameters:
+	/// Execution environment type
+	Environment(ExecutionEnvironment),
```

Comment:
Do we really care what kind of wasm executor we are going to use? I don't think so. What is with other implementations that don't use wasmtime? That can not use wasmtime because they are written in another language?

Reply:
But suppose we use two different versions of Wasmtime. In that case, they are two different execution environments (as they may produce different artifacts from the same Wasm input, and they may enforce limits differently).

Reply:
I'm still not convinced that this is something we should strive for. We should strive for having both versions/environments having the same semantics and leading to the same results. If we see that this doesn't work, we can still bring back this variant in the future when we need to do some migration.